rfcs/OptInFeatures.md

```diff
@@ -0,0 +1,152 @@
+```
+
+The `optIn` directive can then be used in the schema. For an example, to signal an experimental field:
+
+```graphql
+type Session {
+  id: ID!
+  title: String!
+  # [...]
+  startInstant: Instant @optIn(feature: "experimentalInstantApi")
+  endInstant: Instant @optIn(feature: "experimentalInstantApi")
+}
+```
+
+### Introspection
+
+> This section is a proposal based on the current introspection mechanism. A more global mechanism (see [#300](https://github.com/graphql/graphql-spec/issues/300)) would make it obsolete
+
+`@optIn` features should be hidden from introspection by default and include if `includeOptIn` contains the given feature:
+
+```graphql
+type __Type {
+  kind: __TypeKind!
+  name: String
+
+  # [...] other fields omitted for clarity
+
+  # includeOptIn is a list of features to include
+  fields(includeDeprecated: Boolean = false, includeOptIn: [String!]): [__Field!]  
+}
+```
+
+Tools can get a list of `@optIn` features required to use a field (or input field, argument, enum value) using `requiresOptIn`:
+```graphql
+type __Field {
+  name: String!
+  isDeprecated: Boolean!
+
+  # [...] other fields omitted for clarity
+
+  # list of @optIn features required to use this field
+  requiresOptIn: [String!]
+  args(includeDeprecated: Boolean = false, includeOptIn: [String!]): [__InputValue!]!
+}
+```
+
+A given field is included in introspection results if all the conditions are satisfied. In pseudo code, if the following condition is true:
+
+```
+includeOptIn.containsAll(field.requiresOptIn) && (includeDeprecated || !field.isDeprecated)
+```
+
+### Validation
+
+Similarly to [deprecation](https://spec.graphql.org/draft/#sel-FAHnBZNCAACCwDqvK), the `@optIn` directive must not appear on required (non-null without a default) arguments or input object field definitions.
+
+In other words, `@optIn`  arguments or input fields, must be either nullable or have a default value.
+
+## 🗳️ Alternate solutions
+
+### `@experimental` directive
+
+```graphql
+# Indicates that the given field or enum value is still experimental and might be changed 
+# in a backward incompatible manner
+directive @experimental(
+  reason: String! = "Experimental"
+) on FIELD_DEFINITION | ARGUMENT_DEFINITION | INPUT_FIELD_DEFINITION | ENUM_VALUE
+```
+
+Pros:
+* simple
+* symmetrical with `@deprecated`
+
+Cons:
+* doesn't account for opt-in requirements that are not experimental 
+* makes it harder to group by features. `reason` could be used for this but it is less explicit than `feature`
+
+### marker directives
+
+```graphql
+directive @optIn
+```
+
+The user then define their own directives:
+
+```graphql
+# optIn usage defines @experimentalDeploymentApi as an opt-in marker
+directive @experimentalDeploymentApi @optin
+
+type Query {
+  deployment: Deployment @experimentalDeploymentApi
+}
+```
+
+Pros:
+* gives more control to the user about the directive used
+* has more type information
+
+Cons:
+* more complex
+* requires a grammar change
```

Comment:
> Is my understanding correct?

100% correct. I took the liberty of reusing your wording as it's much clearer 👌 and pushed it to the proposal, I hope you don't mind?

> Do clients do something to consume this? Or is it just documenting?

Simply documenting.  See also [my other comment](https://github.com/graphql/graphql-wg/pull/1006#discussion_r889565943). 

> It's not clear what extra information would be useful to capture. Some examples might help clarify

To me, the main advantage is that tools can get the list of opt-in features by scanning the directive definitions. They don't have to scan each field and infer unique feature names. So it's a bit easier on the tooling side.

As a side effect, it also allows users to add arguments to their directives (your `WorkspaceKind` above) although at this point, I'm not sure at all there are use cases for this. 

> This appears to hinder the ability to do out-of-band feature opt-ins (HTTP headers) because you don't have an identifier to use. E.g.

You can always use directive name. You example would actually work

```
headers: {
  "X-ExperimentalApi": "experimentalDeploymentApi"
}
```